python/lsst/daf/butler/core/dimensions/_dataCoordinateIterable.py

```diff
@@ -0,0 +1,764 @@
+
+    __slots__ = ("_dataId",)
+
+    def __iter__(self) -> Iterator[DataCoordinate]:
+        yield self._dataId
+
+    def __len__(self) -> int:
+        return 1
+
+    def __contains__(self, key: Any) -> bool:
+        if isinstance(key, DataCoordinate):
+            return key == self._dataId
+        else:
+            return False
+
+    @property
+    def graph(self) -> DimensionGraph:
+        # Docstring inherited from DataCoordinateIterable.
+        return self._dataId.graph
+
+    def hasFull(self) -> bool:
+        # Docstring inherited from DataCoordinateIterable.
+        return self._dataId.hasFull()
+
+    def hasRecords(self) -> bool:
+        # Docstring inherited from DataCoordinateIterable.
+        return self._dataId.hasRecords()
+
+    def subset(self, graph: DimensionGraph) -> _ScalarDataCoordinateIterable:
+        # Docstring inherited from DataCoordinateIterable.
+        return _ScalarDataCoordinateIterable(self._dataId.subset(graph))
+
+
+class _DataCoordinateCollectionBase(DataCoordinateIterable):
+    """A partial `DataCoordinateIterable` implementation that is backed by a
+    native Python collection.
+
+    This class is intended only to be used as an intermediate base class for
+    `DataCoordinateIterables` that assume a more specific type of collection
+    and can hence make more informed choices for how to implement some methods.
+
+    Parameters
+    ----------
+    dataIds : `collections.abc.Collection` [ `DataCoordinate` ]
+         A collection of `DataCoordinate` instances, with dimensions equal to
+        ``graph``.
+    graph : `DimensionGraph`
+        Dimensions identified by all data IDs in the set.
+    hasFull : `bool`, optional
+        If `True`, the caller guarantees that `DataCoordinate.hasFull` returns
+        `True` for all given data IDs.  If `False`, no such guarantee is made,
+        and `hasFull` will always return `False`.  If `None` (default),
+        `hasFull` will be computed from the given data IDs, immediately if
+        ``check`` is `True`, or on first use if ``check`` is `False`.
+    hasRecords : `bool`, optional
+        If `True`, the caller guarantees that `DataCoordinate.hasRecords`
+        returns `True` for all given data IDs.  If `False`, no such guarantee
+        is made and `hasRecords` will always return `False`.  If `None`
+        (default), `hasRecords` will be computed from the given data IDs,
+        immediately if ``check`` is `True`, or on first use if ``check`` is
+        `False`.
+    check: `bool`, optional
+        If `True` (default) check that all data IDs are consistent with the
+        given ``graph`` and state flags at construction.  If `False`, no
+        checking will occur.
+    """
+    def __init__(self, dataIds: Collection[DataCoordinate], graph: DimensionGraph, *,
+                 hasFull: Optional[bool] = None, hasRecords: Optional[bool] = None,
+                 check: bool = True):
+        self._dataIds = dataIds
```

Comment:
I thought about that a lot.  Since sometimes these will be pretty big collections, I preferred trusting the user to logically transfer ownership over making a (shallow) copy whether one was needed or not.  But maybe that's just premature optimization.

Reply:
maybe update the docstring then to say that the caller should not modify it